### PR TITLE
[FX] Dynamic Shapes Support

### PR DESCRIPTION
**Pull request by anzr299**
### Changes

Modify NNCF Graph Builder for FX backend to correctly get and insert the dynamic shapes into NNCFGraph

### Reason for changes

To support quantization of Torch FX models exported with dynamic shapes

### Tests

test is added to `tests/torch/fx/test_models.py` in test_quantized_models(). Currently only the synthetic transformer is tested because torch.export.dynamic_shapes.Dim.DYNAMIC is not supported in pytorch but is supported in upcoming releases. https://pytorch.org/tutorials/intermediate/torch_export_tutorial.html#constraints-dynamic-shapes

`test_dynamic_edge()` is also added in `tests/torch/fx/test_models.py` to check that the tensor shape in NNCF Graph edge has values only of type int or str and not SymInt.

### CODE REVIEW

File: tests/torch/fx/test_models.py [hunk 1]
```diff
@@ -171,18 +176,29 @@ def test_model(test_case: ModelCase):
 )
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_models.py:178]: Looks like some test cases are duplicated: non synthetic transformer models still are being checked twice with the static shapes. Please refactor the test to prevent redundant cases and please try to avoid pytest.skip call as well

**Review by daniil-lyakhov — CHANGES_REQUESTED** [review]

File: tests/torch/fx/test_models.py (hunk 1 above)
Comment (anzr299) [reply tests/torch/fx/test_models.py:178]: Would you suggest I rather remove the graph tests and only retain `test_dynamic_edge()` test since most of the transformer models in this test are not correctly traced with dynamic shapes

**Comment by daniil-lyakhov** [comment]
NNCFGraphEdge and NNCF algorithms expect the edge shape to be a `List[int]` https://github.com/openvinotoolkit/nncf/blob/develop/nncf/common/graph/graph.py. The use of an str as a shape dimension could lead to unexpected errors in the algorithms: what would happened if a channel dim will be "s1" for a per-channel quantization?

Could you please show an example with a quantization of a dynamic model? Perhaps we can cover the case with less drastic change (using -1 as a shape dim, for example)

**Comment by anzr299** [comment]
> NNCFGraphEdge and NNCF algorithms expect the edge shape to be a `List[int]` https://github.com/openvinotoolkit/nncf/blob/develop/nncf/common/graph/graph.py. The use of an str as a shape dimension could lead to unexpected errors in the algorithms: what would happened if a channel dim will be "s1" for a per-channel quantization?
> 
> Could you please show an example with a quantization of a dynamic model? Perhaps we can cover the case with less drastic change (using -1 as a shape dim, for example)
I can add an exception for the case where the in_channel has a non-static .
using integer such as -1 could cause more complex dynamic shape definition (for example `s0//2*3+s1`) in int to resolve to another value. Would you recommend simply catching the known cases? or to brainstorm on other ways to represent the dynamic shapes

**Comment by anzr299** [comment]
1. I have tried with -1 as dynamic shape. It seems to work fine and pass all tests. I also tested with a random constant shape  for all the tensor shape values except for channel shape. For example, [1,3,224,224] -> [1000, 3, 1000, 1000] where the random constant is 1000. This also did not cause any issues for torch FX tests. I tested this for all the shapes in every model. So it looks like there is no issue with the tensor shapes.
2. I am looking into ways to obtain the maximum value. I did find [this](https://pytorch.org/docs/stable/torch.compiler_dynamic_shapes.html#the-guard-model) which talks about how torch utilizes the `hint` value of symbolic data types. I think this could also be a good replacement for dynamic shape in nncf graph.

File: tests/torch/fx/test_models.py (hunk 1 above)
Comment (daniil-lyakhov) [reply tests/torch/fx/test_models.py:178]: Yes, that would solve the duplicates. You can create a helper function in which you can keep the duplicated body of the test

File: nncf/experimental/torch/fx/nncf_graph_builder.py [hunk 2]
```diff
@@ -197,6 +197,9 @@ def get_edge_params(
                 tensor = source_node.meta["val"]
             if isinstance(tensor, torch.Tensor):
                 tensor_shape = tuple(tensor.shape)
+                tensor_shape = tuple(i.node.hint if isinstance(i, torch.SymInt) else i for i in tensor_shape)
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/fx/nncf_graph_builder.py:200]:
```suggestion
                tensor_shape = tuple(i.node.hint if isinstance(i, torch.SymInt) else i for i in tensor.shape)
```

File: tests/torch/fx/test_models.py [hunk 3]
```diff
@@ -208,6 +235,28 @@ def transform_fn(data_item):
...
+        )
+    ]
+    fx_model = get_torch_fx_model(model, ex_inputs, dynamic_shapes=dynamic_shapes)
+    nncf_graph = GraphConverter.create_nncf_graph(fx_model)
+
+    for edge in nncf_graph.get_all_edges():
+        edge_shape = edge.tensor_shape
+        assert isinstance(edge_shape, tuple)
+        for dim in edge_shape:
+            assert isinstance(dim, (int, str))
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_models.py:256]:
```suggestion
            assert isinstance(dim, int)
```

File: tests/torch/fx/test_models.py [hunk 4]
```diff
@@ -157,32 +166,46 @@ def test_model(test_case: ModelCase):
             (250, 250),
             (149, 149),
         ],
+        [True, False, True, True],
     ),
     (
         ModelCase(partial(ShortTransformer, 5, 10), "synthetic_transformer", [5]),
         {"model_type": nncf.ModelType.TRANSFORMER},
         [(4, 4), (2, 2)],
+        [True],
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_models.py:175]: Can we build dynamic shapes directly here, without bool -> dim conversion?
```suggestion
        [Dim.Auto],
```

**Review by daniil-lyakhov — COMMENTED** [review]
Can we reuse existing reference dot files? AFAIK they look identical

File: tests/torch/fx/test_models.py (hunk 4 above)
Comment (anzr299) [reply tests/torch/fx/test_models.py:175]: Yes, that is possible. I used bool since it was simpler type and to avoid repetition. I will change it as suggested

File: nncf/experimental/torch/fx/nncf_graph_builder.py (hunk 2 above)
Comment (anzr299) [reply nncf/experimental/torch/fx/nncf_graph_builder.py]: Done

File: tests/torch/fx/test_models.py (hunk 3 above)
Comment (anzr299) [reply tests/torch/fx/test_models.py]: Done

File: tests/torch/fx/test_models.py (hunk 4 above)
Comment (anzr299) [reply tests/torch/fx/test_models.py]: Done

File: tests/torch/fx/test_statistics_aggregator.py [hunk 5]
```diff
@@ -59,7 +60,15 @@ def get_fast_bias_correction_algo_backend_cls(self) -> Type[FXFastBiasCorrection
...
         conv_w = self.dataset_samples_to_conv_w(np.array(sample))
-        return get_torch_fx_model(IdentityConv(conv_w), torch.ones(INPUT_SHAPE))
+        dynamic_shapes = [
+            (
+                Dim.AUTO,
+                Dim.STATIC,
+                Dim.AUTO,
+                Dim.AUTO,
+            )
+        ]
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_statistics_aggregator.py:70]: What's the reason behind this change?

File: tests/torch/fx/test_models.py [hunk 6]
```diff
@@ -208,6 +235,28 @@ def transform_fn(data_item):
...
+    ]
+    fx_model = get_torch_fx_model(model, ex_inputs, dynamic_shapes=dynamic_shapes)
+    nncf_graph = GraphConverter.create_nncf_graph(fx_model)
+
+    for edge in nncf_graph.get_all_edges():
+        edge_shape = edge.tensor_shape
+        assert isinstance(edge_shape, tuple)
+        for dim in edge_shape:
+            assert isinstance(dim, int)
+            assert not isinstance(dim, torch.SymInt)
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_models.py:257]:
```suggestion
        assert all(isinstance(dim, int) for dim in edge_shape)
```

**Review by daniil-lyakhov — APPROVED** [review]

File: nncf/experimental/torch/fx/nncf_graph_builder.py [hunk 7]
```diff
@@ -196,7 +196,9 @@ def get_edge_params(
             else:
                 tensor = source_node.meta["val"]
             if isinstance(tensor, torch.Tensor):
-                tensor_shape = tuple(tensor.shape)
+                tensor_shape = tuple(-1 if isinstance(i, torch.SymInt) else i for i in tensor.shape)
+            if isinstance(tensor, torch.SymInt):
```
Comment (AlexanderDokuchaev) [inline nncf/experimental/torch/fx/nncf_graph_builder.py:200]:
```suggestion
            elif isinstance(tensor, torch.SymInt):
```

**Review by AlexanderDokuchaev — CHANGES_REQUESTED** [review]

**Review by AlexanderDokuchaev — APPROVED** [review]
